eglot.el

```
@@ -1169,15 +1217,22 @@ If optional MARKER, return a marker instead"
   "URIfy PATH."
   (url-hexify-string
    (concat "file://" (if (eq system-type 'windows-nt) "/")
-           (directory-file-name (file-truename path)))
+           (directory-file-name (file-local-name (file-truename path))))
    url-path-allowed-chars))
 
-(defun eglot--uri-to-path (uri)
-  "Convert URI to a file path."
+(defun eglot--uri-to-path (uri server)
+  "Convert URI to a file path on SERVER.
```

I've mentioned this is a previous comment.  Can you explain why using `current-server-or-lose` is a bad idea here?

---

This was just a gut feeling of mine. It came down to a few things:

1. The `uri-to-path` function seems pretty low level to me, and thus can't really know all the contexts in which it's going to be called. Since it requires the server information to perform its duty, it felt like it should also be a required argument.
2. `current-server-or-lose` is basically a global variable, and while useful, limiting its uses to higher-level code makes it more tractable.
3. Lastly, aesthetics. I find the above easier to read and reason about than `(defun eglot--uri-to-path (uri &optional (server (eglot--current-server-or-lose)))…)`

Once I tracked through the places in the code where this function was being called, it turned out to be pretty easy to supply the `server` argument from places where it was passed in via server methods. The remaining places where I had to use `current-server-or-lose` as its value, `current-server-or-lose` was already being used in that section of code (like the `xref` stuff). So that made me feel more confident about the decision to add the new argument.

---

> current-server-or-lose is basically a global variable, and while useful, limiting its uses to higher-level code makes it more tractable.

It's a function it's implemented with a special variable, not exactly a global one. Crucially, the var is not writable per the protocol, unless you violate the protocol. In CL, and in Emacs you have lots of these, just look at CL's `*print-...` stuff.  They influence "lower-level" functions deeply too.

> 3. Lastly, aesthetics. 

You don't have to use an optional argument.  And you needn't give it a default in the `defun`.  And it doesn't work with Elisp defun anyway (you need `cl-defun`).

I'd rather we don't thread the server argument down to every nook and cranny if we can avoid even thinking about it.  Because doing the same thing repeatedly is a maintenance risk, in general.  Bottom line is we want this function to fail loudly if the server isn't available.

---

It hadn't occurred to me that I didn't need an optional argument, but of course you are right. I still disagree, as I think the other two points stand, but it's not a big deal to me.

I'll make this change and have `eglot--uri-to-path` be single argument again.

---

thank you.  I do find it curious why you think of `eglot--uri-to-path` as "low level". Shouldn't "low-level" function be able to see the dynamic environment?  Or do you want to call it from ouside that environment.  Say, in a debug setting.

By the way, if the server is indeed **optional**, then the correct choice is `eglot--current-server`, **not** `eglot--current-server-or-lose`.  By doing so, you retain the possibility to call `eglot--uri-to-path` from serverless contexts.

---

Again, this is all just gut. I find it hard to express exactly why I think it's better to have it be explicit. I look at `uri-to-path` as something that does a strict mapping from a URI to a path, a pure function who's inputs soley determine the outputs. Given that view, relying on external state isn't possible.

Now, I confess that I'm pretty fond of pure functions for lots of reasons, and so I try to use them as much as I think is reasonable. An extension of this is to try and limit global context as much as is reasonable.

So to me, it's a trade-off: is it worse to thread the argument through the code, or worse to rely on external context? I felt in this case, the latter, because the function *feels* like a pure function to me, and there wasn't a huge amount of threading necessary, and where it was, it was easy to accomplish.

But this is just a judgement call, and reasonable people can be expected to disagree. I can certainly see why you'd rather rely on using the external state.

---

Heh.  Interesting.  In my view, the function's purity is broken as soon as its main job (converting an URI) is now dependent on an external condition that you pass into the function. You're just dressing up the "impurity" in other ways.  What's the difference between relying on a bunch of arguments threaded down and relying on a bunch of dynamically scoped environment variables?  Just aesthetics.  But I can certainly see your point too.  I've often taken the opposite stance, and sometimes I will thread arguments.